### PR TITLE
Refactoring the node server and client.

### PR DESCRIPTION
**Pull request by raduom**
1) Hoisting of ChainSyncServer onto IO.
2) Use MTL for writing the state transition code.
3) Remove a bunch of where blocks, make variable provenance clearer.

### CODE REVIEW

File: plutus-scb/src/Cardano/Protocol/Socket/Server.hs [hunk 1]
```diff
@@ -32,173 +41,392 @@ import           Ouroboros.Network.Protocol.Handshake.Version
...
+    ( MonadReader InternalState m
+    , MonadIO m )
+ => LocalChannel
+ -> m (Either (ServerStNext Block Tip m ())
+              (m (ServerStNext Block Tip m ())))
+nextState localChannel@(LocalChannel channel) = do
+    chainState <- ask <&> isState
+    tip <- head . view Chain.chainNewestFirst <$>
+           liftIO (readMVar chainState)
+    (liftIO . atomically $ tryReadTChan channel) >>= \case
```
Comment (michaelpj) [inline plutus-scb/src/Cardano/Protocol/Socket/Server.hs:199]:
```suggestion
    maybeBlock <- liftIO . atomically $ tryReadTChan channel
    case maybeBlock of
```

File: plutus-scb/src/Cardano/Protocol/Socket/Server.hs [hunk 2]
```diff
@@ -32,173 +41,392 @@ import           Ouroboros.Network.Protocol.Handshake.Version
...
 
-{-| This function starts a blockchain simulator as seen through a
-    Cardano node.
+type CommandQueue = TQueue (Either ServerCommand ServerResponse)
+type Error a = Either String a
 
-    It received the path to the communication socket, and the server
-    application state, then waits for requests using the `ChainSync`
-    and `LocalTxSubmission` protocols.
+{- We clone the original channel for each connected client. We use
```
[the remark this answers is not in the document]
Comment (raduom) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:53]: I tried to use {- .. -} for comments on functions which are internal and I don't want to be exposed. Is that a bad idea?

File: plutus-scb/src/Cardano/Protocol/Socket/Server.hs [hunk 3]
```diff
@@ -32,173 +41,392 @@ import           Ouroboros.Network.Protocol.Handshake.Version
...
 
 import           Ledger                                              (Block, Slot (..), Tx (..))
-import           Wallet.Emulator.Chain
+import           Wallet.Emulator.Chain                               (ChainEvent (..), ChainState (..))
+import qualified Wallet.Emulator.Chain                               as Chain
 
-{-| This function starts a blockchain simulator as seen through a
-    Cardano node.
+type CommandQueue = TQueue (Either ServerCommand ServerResponse)
+type Error a = Either String a
```
[the remark this answers is not in the document]
Comment (raduom) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:51]: Isn't there an alternative to defining my own `Error` type?

[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:51]: Nothing wrong with defining your own error type!

File: plutus-scb/src/Cardano/Protocol/Socket/Server.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:53]: I thought this was actually a useful comment, hence the suggestion.

File: plutus-scb/src/Cardano/Protocol/Socket/Server.hs (hunk 1 above)
Comment (raduom) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:199]: Is there a reason why you don't like this usage of `LambdaCase`?

Comment (michaelpj) [reply plutus-scb/src/Cardano/Protocol/Socket/Server.hs:199]: It's not that I don't like, it's `>>=` with complex expressions on either side. I just find it usually hard to read.

**Pull request merged**